src/linkediting.js

```diff
@@ -61,13 +77,100 @@ export default class LinkEditing extends Plugin {
 		editor.commands.add( 'link', new LinkCommand( editor ) );
 		editor.commands.add( 'unlink', new UnlinkCommand( editor ) );
 
+		const linkDecorators = editor.config.get( 'link.decorators' ) || [];
+
+		this._enableAutomaticDecorators( linkDecorators.filter( item => item.mode === DECORATOR_AUTOMATIC ) );
+		this._enableManualDecorators( linkDecorators.filter( item => item.mode === DECORATOR_MANUAL ) );
+
 		// Enable two-step caret movement for `linkHref` attribute.
 		bindTwoStepCaretToAttribute( editor.editing.view, editor.model, this, 'linkHref' );
 
 		// Setup highlight over selected link.
 		this._setupLinkHighlight();
 	}
 
+	/**
+	 * Processes an array of configured {@link module:link/link~LinkDecoratorAutomaticOption automatic decorators}
+	 * and registers a {@link module:engine/conversion/downcastdispatcher~DowncastDispatcher downcast dispatcher}
+	 * for each one of them. Downcast dispatchers are obtained using the
+	 * {@link module:link/utils~AutomaticDecorators#getDispatcher} method.
+	 *
+	 * **Note**: This method also activates the automatic external link decorator if enabled via
+	 * {@link module:link/link~LinkConfig#targetDecorator `config.link.targetDecorator`}.
+	 *
+	 * @private
+	 * @param {Array.<module:link/link~LinkDecoratorAutomaticOption>} automaticDecoratorDefinitions
+	 */
+	_enableAutomaticDecorators( automaticDecoratorDefinitions ) {
+		const editor = this.editor;
+		const automaticDecorators = new AutomaticDecorators();
+
+		// Adds default decorator for external links.
+		if ( editor.config.get( 'link.targetDecorator' ) ) {
+			automaticDecorators.add( {
+				mode: DECORATOR_AUTOMATIC,
+				callback: url => EXTERNAL_LINKS_REGEXP.test( url ),
+				attributes: {
+					target: '_blank',
+					rel: 'noopener noreferrer'
+				}
+			} );
+		}
+
+		automaticDecorators.add( automaticDecoratorDefinitions );
+
+		if ( automaticDecorators.length ) {
+			editor.conversion.for( 'downcast' ).add( automaticDecorators.getDispatcher() );
+		}
+	}
+
+	/**
+	 * Processes an array of configured {@link module:link/link~LinkDecoratorManualOption manual decorators}
+	 * and transforms them into {@link module:link/utils~ManualDecorator} instances and stores them in the
+	 * {@link module:link/linkcommand~LinkCommand#manualDecorators} collection (a model for manual decorators state).
+	 *
+	 * Also registers an {@link module:engine/conversion/downcasthelpers~DowncastHelpers#attributeToElement attributeToElement}
+	 * converter for each manual decorator and extends the {@link module:engine/model/schema~Schema model's schema}
+	 * with adequate model attributes.
+	 *
+	 * @private
+	 * @param {Array.<module:link/link~LinkDecoratorManualOption>} manualDecoratorDefinitions
+	 */
+	_enableManualDecorators( manualDecoratorDefinitions ) {
```

Comment:
The function `_enableManualDecorators` accepts `manualDecoratorDefinitions` which are `LinkDecoratorManualOption`. It's counterintuitive.

We should either rename the `LinkDecoratorManualOption` type to `LinkDecoratorManualDefinition` (which makes sense to me) or use `manualDecoratorOptions` as an argument.